src/os/apple/register.rs

```diff
@@ -139,27 +148,85 @@ impl Drop for ServiceRef {
 
 impl RegisteredDnsService {}
 
-// In order to signal the blocked thread, we close its socket to unblock it
+// Signal the background thread via an atomic flag
 impl Drop for RegisteredDnsService {
     fn drop(&mut self) {
-        unsafe {
-            trace!("Closing socket to signal service cleanup...");
-            libc::close(self.socket);
-        }
+        trace!("Signaling thread to exit...");
+        self.shutdown_flag.store(true, Ordering::Release);
     }
 }
-fn run_thread(service: ServiceRef) {
-    std::thread::spawn(move || loop {
-        unsafe {
-            trace!("Processing...");
-            let r = DNSServiceProcessResult(service.raw);
-            if r != kDNSServiceErr_NoError {
-                error!("Error processing: {}, exiting thread", r);
+fn run_thread_with_poll(service: ServiceRef, shutdown_flag: Arc<AtomicBool>) {
+    let socket = unsafe { DNSServiceRefSockFD(service.raw) };
+
+    std::thread::spawn(move || {
+        loop {
+            // Check a shutdown flag
+            if shutdown_flag.load(Ordering::Acquire) {
+                trace!("Shutdown requested, exiting DNS-SD processing thread");
                 break;
             }
+
+            unsafe {
+                let mut poll_fd = libc::pollfd {
+                    fd: socket,
+                    events: libc::POLLIN,
+                    revents: 0,
+                };
```

Comment:
I assume we don't need any kind of cleanup for `libc::pollfd` (e.g. deallocation etc) – just because we're inside `unsafe` promise scope.

Reply:
yeah it's just a struct on the stack, and no cleanup that I'm aware of that's needed